### PR TITLE
support convlstm

### PR DESCRIPTION
**Pull request by dding3**
## What changes were proposed in this pull request?
1. support convlstm 
2. add related ut

## How was this patch tested?
new ut and manual test

### CODE REVIEW

**Comment by dding3** [comment]
Jekins pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala [hunk 1]
```diff
@@ -71,7 +71,7 @@ abstract class Cell[T : ClassTag](
    * @param size batchSize
    * @return
    */
-  def hidResize(hidden: Activity, size: Int): Activity = {
+  def hidResize(hidden: Activity, size: Int, rows: Int = 1, columns: Int = 1): Activity = {
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala]: If overriding preTopology in Cell, the rows and columns can be removed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ConvLSTMPeephole.scala [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    hiddensShape = Array(hiddenSize, hiddenSize),
+    regularizers = Array(wRegularizer, uRegularizer, bRegularizer)
+  ) {
+  var inputGate: Sequential[T] = _
+  var forgetGate: Sequential[T] = _
+  var outputGate: Sequential[T] = _
+  var hiddenLayer: Sequential[T] = _
+  var cellLayer: Sequential[T] = _
+  override var cell: AbstractModule[Activity, Activity, T] = buildConvLSTM()
+
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ConvLSTMPeephole.scala]:
```
override def preTopology: AbstractModule[Activity, Activity, T] =
         Sequential()
      .add(Contiguous())
      .add(ParallelTable()
          .add(SpatialConvolution(inputSize, hiddenSize, kernelI, kernelI,
                stride, stride, kernelI/2, kernelI/2, wRegularizer = wRegularizer,
                bRegularizer = bRegularizer))
        .add(SpatialConvolution(inputSize, hiddenSize, kernelI, kernelI,
          stride, stride, kernelI/2, kernelI/2, wRegularizer = wRegularizer,
          bRegularizer = bRegularizer))
        .add(SpatialConvolution(inputSize, hiddenSize, kernelI, kernelI,
          stride, stride, kernelI/2, kernelI/2, wRegularizer = wRegularizer,
          bRegularizer = bRegularizer))
        .add(SpatialConvolution(inputSize, hiddenSize, kernelI, kernelI,
          stride, stride, kernelI/2, kernelI/2, wRegularizer = wRegularizer,
          bRegularizer = bRegularizer)))
```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ConvLSTMPeephole.scala [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  def buildGate(): Sequential[T] = {
+    val i2g = Sequential()
+      .add(Contiguous())
+      .add(SpatialConvolution(inputSize, hiddenSize, kernelI, kernelI,
+      stride, stride, kernelI/2, kernelI/2, wRegularizer = wRegularizer,
+        bRegularizer = bRegularizer))
+
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ConvLSTMPeephole.scala]: val i2g = Narrow(dimension, offset, length)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ConvLSTMPeephole.scala [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  def buildHidden(): Sequential[T] = {
+    val hidden = Sequential()
+      .add(NarrowTable(1, 2))
+
+    val i2h = Sequential()
+      .add(Contiguous())
+      .add(SpatialConvolution(inputSize, hiddenSize, kernelI, kernelI,
+      stride, stride, kernelI/2, kernelI/2, wRegularizer = wRegularizer,
+      bRegularizer = bRegularizer))
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ConvLSTMPeephole.scala]: i2h = Narrow(dimension, offset, length)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 5]
```diff
@@ -169,18 +170,23 @@ class Recurrent[T : ClassTag]()
...
-    extend(times, batchSize, hiddenSize)
+    if (input.dim() == 3) {
+      output.resize(batchSize, times, hiddenSize)
+      // Clone N modules along the sequence dimension.
+      extend(times, batchSize, hiddenSize)
+    } else {
+      output.resize(batchSize, times, hiddenSize, input.size(4), input.size(5))
+      // Clone N modules along the sequence dimension.
+      extend(times, batchSize, hiddenSize, input.size(4), input.size(5))
+    }
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: 5D input?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala [hunk 6]
```diff
@@ -84,7 +84,7 @@ abstract class Cell[T : ClassTag](
...
     if (hidden == null) {
       if (hiddensShape.length == 1) {
         hidResize(Tensor[T](), size)
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala]: `hidResize(Tensor[T](), size, rows, columns)`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 7]
```diff
@@ -204,11 +205,16 @@ class Recurrent[T : ClassTag]()
     batchSize = outputCell.size(batchDim)
     times = outputCell.size(timeDim)
 
-    val hiddenSize = topology.hiddensShape(0)
-    output.resize(batchSize, times, hiddenSize)
-
-    // Clone N modules along the sequence dimension.
-    extend(times, batchSize, hiddenSize)
+    val hiddenSize = modules.last.asInstanceOf[Cell[T]].hiddensShape(0)
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: `topology = modules.last.asInstanceOf[Cell[T]]`
They are the same

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala [hunk 8]
```diff
@@ -204,11 +205,16 @@ class Recurrent[T : ClassTag]()
     batchSize = outputCell.size(batchDim)
     times = outputCell.size(timeDim)
 
-    val hiddenSize = topology.hiddensShape(0)
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala]: can keep this line

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala [hunk 9]
```diff
@@ -106,9 +106,16 @@ abstract class Cell[T : ClassTag](
         require(hidden.isInstanceOf[Table],
           "Cell: hidden should be a Table")
         var i = 1
-        while (i <= hidden.toTable.length()) {
-          hidden.toTable[Tensor[T]](i).resize(size, hiddensShape(i - 1))
-          i += 1
+        if (rows== 1 && columns==1) {
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala]: `rows == 1 && columns == 1`

Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Cell.scala]: add space ..

**Comment by JerryYanWan** [comment]
LGTM

**Pull request merged**